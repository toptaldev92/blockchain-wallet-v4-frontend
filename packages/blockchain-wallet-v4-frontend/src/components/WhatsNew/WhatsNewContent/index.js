import React from 'react'
import { FormattedMessage } from 'react-intl'

const Announcements = [
  {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title12' defaultMessage={'Import Funds from MyEtherWallet'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc12' defaultMessage={'You can now import your ether from your MyEtherWallet backup directky into your Blockchain wallet.'}/>,
    date: 'February 10, 2018',
    // link: <a href="blockchain.info/settings/info">Learn More</a>
    //if i have a link, follow link. Otherwise, set an onClick
  }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title11' defaultMessage={'New Bitcoin Cash Address Format'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc11' defaultMessage={'Bitcoin Cash addresses now follow the CashAddr format.'} />,
      date: 'January 18, 2018',
      link: '/settings/info'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title10' defaultMessage={'Bitcoin Cash'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc10' defaultMessage={'If you had bitcoin in your wallet before August 1, 2017, you can now access Bitcoin Cash via Settings > General.'} />,
      date: 'October 3, 2017',
      link: '/settings/info'
    }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title9' defaultMessage={'Exchange Assets'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc9' defaultMessage={'You can exchange between bitcoin, ether, and bitcoin cash directly from your Blockchain wallet'} />,
    date: 'August 11, 2017',
    link: '/exchange'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title8' defaultMessage={'Send & Receive Ether'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc8' defaultMessage={'You can now send and receive ether from your Blockchain wallet!'} />,
    date: 'July 24, 2017',
    link: '/eth/transactions'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title7' defaultMessage={'Sell Bitcoin'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc7' defaultMessage={'You can now sell bitcoin directly from your Blockchain wallet!'} />,
    date: 'May 12, 2017',
    link: '/buy-sell'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title6' defaultMessage={'Buy Bitcoin'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc6' defaultMessage={'You can now buy bitcoin with your Blockchain wallet!'} />,
    date: 'December 15, 2016',
    link: '/buy-sell'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title1' defaultMessage={'Dev Themes'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc1' defaultMessage={'You can now customize your Blockchain Wallet with different themes. Head over to Preferences to see your options.'} />,
    date: 'September 26, 2016'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title5' defaultMessage={'Export History'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc5' defaultMessage={'Export the transaction history of your addresses in CSV format'} />,
    date: 'June 21, 2016'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title4' defaultMessage={'What\'s New'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc4' defaultMessage={'We now have a "What\'s New" section where we\'ll let you know about all the changes we\'ve been making to the Blockchain Wallet.'} />,
    date: 'May 20, 2016'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title3' defaultMessage={'Sign and Verify'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc3' defaultMessage={'Prove that you own a bitcoin address by signing a message. Verify to confirm someone else owns an address.'} />,
    date: 'May 2, 2016'
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title2' defaultMessage={'Transfer All'} />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc2' defaultMessage={'Move all funds from your imported addresses to your HD addresses for added privacy and security.'} />,
    date: 'April 25, 2016'
  }
]

export default Announcements
