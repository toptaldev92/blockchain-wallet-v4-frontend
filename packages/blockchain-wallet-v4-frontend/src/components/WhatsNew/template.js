import React from 'react'
import styled from 'styled-components'
import { FormattedMessage } from 'react-intl'

import { Text, TextGroup, ModalHeader, ModalBody } from 'blockchain-info-components'
import PropTypes from 'prop-types'

const Fragment = React.Fragment

const Wrapper = styled.div`
  box-sizing: border-box;
  overflow: none;
`

const ContentWrapper = styled.div`
  overflow-x: hidden;
  overflow-y: auto;
`
const WhatsNew = (props) => {
  const { handleTrayRightToggle } = props

  const announcements = [
    {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title10' defaultMessage={'Bitcoin Cash'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc10' defaultMessage={'If you had bitcoin in your wallet before August 1, 2017, you can now access Bitcoin Cash via Settings > General.'} />,
      date: 'October 3 2017',
      link: '/settings/info'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title9' defaultMessage={'Exchange Assets'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc9' defaultMessage={'You can now exchange your bitcoin for ether and vice versa directly from your Blockchain wallet.'} />,
      date: 'August 11 2017',
      link: '/exchange'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title8' defaultMessage={'Send & Receive Ether'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc8' defaultMessage={'You can now send and receive ether from your Blockchain wallet!'} />,
      date: 'July 24 2017',
      link: '/eth/transactions'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title7' defaultMessage={'Sell Bitcoin'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc7' defaultMessage={'You can now sell bitcoin directly from your Blockchain wallet!'} />,
      date: 'May 12 2017',
      link: '/buy-sell'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title6' defaultMessage={'Buy Bitcoin'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc6' defaultMessage={'You can now buy bitcoin with your Blockchain wallet!'} />,
      date: 'December 15 2016',
      link: '/buy-sell'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title1' defaultMessage={'Dev Themes'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc1' defaultMessage={'You can now customize your Blockchain Wallet with different themes. Head over to Preferences to see your options.'} />,
      date: 'September 26 2016'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title5' defaultMessage={'Export history'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc5' defaultMessage={'Export the transaction history of your addresses in CSV format'} />,
      date: 'June 21 2016'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title4' defaultMessage={'What\'s New'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc4' defaultMessage={'We now have a "What\'s New" section where we\'ll let you know about all the changes we\'ve been making to the Blockchain Wallet.'} />,
      date: 'May 20 2016'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title3' defaultMessage={'Sign and Verify'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc3' defaultMessage={'Prove that you own a bitcoin address by signing a message. Verify to confirm someone else owns an address.'} />,
      date: 'May 2 2016'
    }, {
      title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title2' defaultMessage={'Transfer All'} />,
      desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc2' defaultMessage={'Move all funds from your imported addresses to your HD addresses for added privacy and security.'} />,
      date: 'April 25 2016'
    }
  ]

  return (
    <Fragment>
      <ModalHeader onClose={() => handleTrayRightToggle()}>
        <FormattedMessage id='layouts.wallet.trayright.whatsnew' defaultMessage='Whats New'/>
      </ModalHeader>
      <ModalBody>
        <Wrapper>
          <ContentWrapper>
            {announcements.map((item, i) => {
              const { title, desc, date } = item
              return (
                <Fragment key={i}>
                  <div>
                    <Text size='14px' weight={400}>{title}</Text>
                    <Text color='gray-3' weight={400} size='12px'>{new Date(date).toLocaleDateString()}</Text>
                  </div>
                  <TextGroup inline>
                    <Text size='12px' weight={300}>{desc}</Text>
                  </TextGroup>
                </Fragment>
              )
            })}
          </ContentWrapper>
        </Wrapper>
      </ModalBody>
    </Fragment>
  )
}

WhatsNew.propTypes = {
  handleTrayRightToggle: PropTypes.func.isRequired
}

export default WhatsNew
