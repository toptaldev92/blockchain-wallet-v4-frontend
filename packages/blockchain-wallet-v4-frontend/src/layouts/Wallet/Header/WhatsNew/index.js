import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Link, Icon } from 'blockchain-info-components'

const WhatsNewLink = styled(Link)`
  position: relative;
  margin-top: 4px;

  & > :first-child:hover {
    cursor: pointer;
  }

  ::after {
    opacity: ${props => props.trayRightOpen ? '1' : '0'};
    content: "";
    position: absolute;
    top: 24px;
    left: -2px;
    width: 0;
    height: 0;
    z-index: 3;
    border-left: 11px solid transparent;
    border-right: 11px solid transparent;
    border-bottom: 16px solid ${props => props.theme['white-blue']};
    transition: opacity ${props => props.trayRightOpen ? '.2s' : '0'};
    transition-delay: ${props => props.trayRightOpen ? '.3s' : '.1s'};
  }
`

const WhatsNew = (props) => {
  const { trayRightOpen, handleTrayRightToggle } = props

  return (
    <WhatsNewLink trayRightOpen={trayRightOpen} onClick={() => handleTrayRightToggle('whats-new')} size='16px' weight={300} color='white'>
      <Icon name='bell-filled' size='18px' color='white'/>
    </WhatsNewLink>
  )
}

WhatsNew.propTypes = {
  handleTrayRightToggle: PropTypes.func.isRequired,
  trayRightOpen: PropTypes.bool.isRequired
}

export default WhatsNew
