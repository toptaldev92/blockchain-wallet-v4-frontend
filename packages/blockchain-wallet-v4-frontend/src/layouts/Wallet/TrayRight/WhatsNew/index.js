import React from 'react'
import styled from 'styled-components'

import ExchangeByBlockchain from './WhatsNewContent/ExchangeByBlockchain'

const Wrapper = styled.div`
  width: 100%;
  height: 100%;
  padding: 30px;
  box-sizing: border-box;
  overflow-y: auto;
`
const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: auto;
`

const WhatsNew = props => (
  <Wrapper>
    <Container>
      <ExchangeByBlockchain />
    </Container>
  </Wrapper>
)

export default WhatsNew
