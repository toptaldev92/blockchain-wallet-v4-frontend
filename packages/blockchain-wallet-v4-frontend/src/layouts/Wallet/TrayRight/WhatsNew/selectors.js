import Announcements from 'services/WhatsNewService/WhatsNewContent'
import { selectors } from 'data'
import moment from 'moment'

export const getData = state => {
  const lastViewed = (selectors.core.kvStore.whatsNew.getLastViewed(state)).getOrElse(moment().subtract(6, 'months').unix())
  const latestAnnouncements = Announcements.filter(announcement => {
    return moment(announcement.date).isAfter(moment(lastViewed))
  })
  return {
    lastViewed,
    latestAnnouncements
  }
}
