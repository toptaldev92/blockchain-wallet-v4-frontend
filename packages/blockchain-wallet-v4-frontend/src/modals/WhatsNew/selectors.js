import { includes, prop } from 'ramda'
import { selectors } from 'data'
import moment from 'moment'
import React from 'react'

import Borrow from './Content/Borrow'
import TxHistoryDownload from './Content/TxHistoryDownload'
import USDDigital from './Content/USDDigital'

const Announcements = [
  {
    content: <TxHistoryDownload key={0} />,
    date: new Date('March 27, 2020'),
    restrictByCountry: [],
    restrictByUserKyc: []
  },
  {
    content: <Borrow key={1} />,
    date: new Date('March 9, 2020'),
    restrictByCountry: [],
    restrictByUserKyc: []
  },
  {
    content: <USDDigital key={2} />,
    date: new Date('March 9, 2020'),
    restrictByCountry: [],
    restrictByUserKyc: []
  }
]

export const filterAnnouncements = (
  lastViewed,
  userCountry,
  userKycState,
  whatsNewKvStore
) => {
  const isOnRestrictedCountryList = includes(userCountry)
  const isOnRestrictedKycList = includes(userKycState)
  const isRestrictedByKvStore = kvCheckFunc => {
    return kvCheckFunc && kvCheckFunc(whatsNewKvStore)
  }
  const isRestricted = announcement =>
    isOnRestrictedCountryList(prop('restrictByCountry', announcement)) ||
    isOnRestrictedKycList(prop('restrictByUserKyc', announcement))

  const isAvailableToView = announcement => {
    const announcementDate = prop('date', announcement)
    const restrictByKv = prop('restrictByKvStoreCheck', announcement)
    // show new announcement alert if
    // the announcement is less than 30 days old &&
    // last time whats new was viewed is before the announcement date &&
    // it is not restricted by kvStore
    return (
      moment()
        .subtract(30, 'days')
        .isBefore(announcementDate) &&
      moment(lastViewed).isBefore(announcementDate) &&
      !isRestrictedByKvStore(restrictByKv)
    )
  }

  return Announcements.map(announcement => ({
    content: prop('content', announcement),
    restricted: isRestricted(announcement),
    display: true,
    alert: isAvailableToView(announcement)
  }))
}

export const getData = state => {
  const whatsNewKvStore = selectors.core.kvStore.whatsNew
    .getState(state)
    .getOrElse({})
  const lastViewed = selectors.core.kvStore.whatsNew
    .getLastViewed(state)
    .getOrElse(
      moment()
        .subtract(2, 'months')
        .unix()
    )
  const userCountry = selectors.core.settings
    .getCountryCode(state)
    .getOrElse(null)
  const userKycState = selectors.modules.profile
    .getUserKYCState(state)
    .getOrElse(null)

  return {
    announcements: filterAnnouncements(
      lastViewed,
      userCountry,
      userKycState,
      whatsNewKvStore
    ).filter(prop('display'))
  }
}
