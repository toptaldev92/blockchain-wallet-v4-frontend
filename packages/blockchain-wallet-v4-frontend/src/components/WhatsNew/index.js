import React from 'react'
import { connect } from 'react-redux'
import { selectors } from 'data'

import WhatsNew from './template.js'

class WhatsNewContainer extends React.PureComponent {
  render () {
    const { handleTrayRightToggle } = this.props
    return (
      <WhatsNew handleTrayRightToggle={handleTrayRightToggle}/>
    )
  }
}

const mapStateToProps = (state) => ({
  countryCode: selectors.core.settings.getCountryCode(state)
})

export default connect(mapStateToProps, undefined)(WhatsNewContainer)
