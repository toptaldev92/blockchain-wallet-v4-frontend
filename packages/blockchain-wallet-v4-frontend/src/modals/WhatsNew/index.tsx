import { actions } from 'data'
import { bindActionCreators, compose } from 'redux'
import { connect } from 'react-redux'
import { getData } from './selectors'
import { prop } from 'ramda'
import EmptyContent from 'services/WhatsNewService/WhatsNewContent/EmptyContent'
import Flyout, { duration } from 'components/Flyout'
import modalEnhancer from 'providers/ModalEnhancer'
import React from 'react'
import styled from 'styled-components'

type OwnPropsType = {
  position: number,
  close: () => void,
  total: number,
  userClickedOutside: boolean
}

type LinkStatePropsType = {
  announcements: Array<{
    content: React.Component
  }>
}

type LinkDispatchPropsType = {
  kvStoreWhatsNewActions: typeof actions.core.kvStore.whatsNew
}

type Props = OwnPropsType & LinkStatePropsType & LinkDispatchPropsType

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: auto;
`

class WhatsNewContainer extends React.PureComponent<Props> {
  state = { show: false }

  componentDidMount () {
    /* eslint-disable */
    this.setState({ show: true })
    /* eslint-enable */
  }

  componentWillUnmount () {
    this.props.kvStoreWhatsNewActions.updateMetadataWhatsNew(Date.now())
  }

  render () {
    const { show } = this.state
    const { announcements, ...rest } = this.props
    return (
      <Flyout {...rest} onClose={this.props.close} in={show} data-e2e='whatsNewModal'>
        <Container>
          {!prop('length', announcements) ? (
            <EmptyContent />
          ) : (
              announcements.map(prop('content'))
            )}
        </Container>
      </Flyout>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  kvStoreWhatsNewActions: bindActionCreators(
    actions.core.kvStore.whatsNew,
    dispatch
  )
})

const enhance = compose<any>(
  modalEnhancer('WHATS_NEW_MODAL', { transition: duration }),
  connect(
    getData,
    mapDispatchToProps
  )
)

export default enhance(WhatsNewContainer)
