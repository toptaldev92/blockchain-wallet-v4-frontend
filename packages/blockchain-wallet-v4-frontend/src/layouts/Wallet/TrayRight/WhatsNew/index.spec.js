import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import WhatsNew from './index'

jest.mock(
  './WhatsNewContent/ExchangeByBlockchain',
  () => 'exchangeByBlockchain'
)

describe('Whats New', () => {
  it('renders correctly', () => {
    const component = shallow(<WhatsNew />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
