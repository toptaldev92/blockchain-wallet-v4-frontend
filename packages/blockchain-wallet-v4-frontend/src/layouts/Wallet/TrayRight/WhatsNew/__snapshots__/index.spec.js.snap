// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whats New renders correctly 1`] = `
<WhatsNew__Wrapper>
  <WhatsNew__Container>
    <exchangeByBlockchain />
  </WhatsNew__Container>
</WhatsNew__Wrapper>
`;
